src/Zynga/Framework/PgData/V1/PgModel/Reader.hh

```diff
@@ -65,52 +67,51 @@ class Reader implements ReaderInterface {
       // 1) grab a copy of our object to work with.
       $pgModel->stats()->incrementCacheMisses();
       $obj = $pgModel->data()->createRowObjectFromClassName($model);
-      $pk = $obj->getPrimaryKeyTyped();
-      $pk->set($id);
 
-      // 2) Lock the row for update from the db data.
-      $pgModel->cache()->lockRowCache($obj);
-      
-      // 3) Stand up the where statement to get the response we want.
+      // 2) Stand up the where statement to get the response we want.
       $where = new PgWhereClause($pgModel);
       $where->and($obj->getPrimaryKey(), PgWhereOperand::EQUALS, $id);
 
-      // 4) Run the query against the database and dont release any of the lock yet
+      // 3) Run the query against the database and dont release any of the lock yet
       $resultSet = $this->fetchResultSetFromDatabase($model, $where, false);
 
       // 4) We expect a result set of 1 for return.
       if ($resultSet->count() == 1) {
         $row = $resultSet->get(0);
         if ($row instanceof PgRowInterface) {
           // fetchResultSetFromDatabase already writes the row to cache.
-          if($getLocked === false) {
-            $pgModel->cache()->unlockRowCache($row);
-          }
-          
+
           return $row;
         }
       }
-      
-      // 5) If we got more than 1 result,
-      // then release lock for all of them acquired by fetchResultSetFromDatabase
-      if($resultSet->count() > 1) {
-        foreach ($resultSet->toArray() as $resultObj) {
-          if (!$resultObj instanceof PgRowInterface) {
-            continue;
-          }
-          
-          $pgModel->cache()->unlockRowCache($resultObj);
-        }
-        
-        return null;
-      }
 
-      // 6) row not found by pk. Always unlock in this case
-      $pgModel->cache()->unlockRowCache($obj);
       return null;
 
     } catch (Exception $e) {
+      // unlock the result obj even asked for not to unlock when an exception happens
+      $getLocked = false;
       throw $e;
+    } finally {
+      // release all the possible locks
+      if ($pgModel !== null && $resultSet !== null) {
+        $cache = $pgModel->cache();
+        foreach ($resultSet->toArray() as $idx => $resultObj) {
+
+          if (!$resultObj instanceof PgRowInterface) {
+            continue;
+          }
```

Comment:
There was couple of comments here from @Mehul:
 1) if result set size > 1, 
 i. first element will be unlocked based on caller ask($getLocked input flag)  
 ii. unlock remaining all if rs has more.
I'm on the same page and still this loop does the same, just trying to avoid 2 sections of code & additional conditions. I've adjusted ($idx == 0) condition lil bit and I think it covers well for part i.

2. $obj - refer the comment added at line 68. I think we don't required that anymore, pls check if i'm missing anything

Reply:
1. if result set size > 1, we return null and don't return the first index in the result set. In that case, we should unlock all rows and not skip first element based on $getLocked flag. 

2. I responded to your comment above. Also fetchSingleRowFromDataCache will grab a lock if $getLocked is true. In that case, its totally possible to get a lock and then fail with an exception on get from cache. When that happens resultSet will be null. 

We need a good way to differentiate between finally called on an exception vs finally called if everything was executed properly. 

Here is what i am thinking:
1) Application code locks a row directly for id = 1;
2) Application code calls getByPk on id = 1 & getLocked = false. 

In that case, getByPk() should only unlock if there is an exception but not when everything was executed properly

Reply:
yeah, I think I missed a tiny case earlier when reseltset size > 1 or exception case, thanks for correcting me.
see this new change covers both cases.

Reply:
The changes makes sense. 

The only other case that we should also cover are:
1) when $fetchSingleRowFromDataCache grabs the lock but throws an exception while trying to read. This is called on line 56. In that case $resultSet will be null.

Reply:
good catch, this is another edge case. I think it makes sense to handle that in fetchSingleRowFromDataCache() method itself (on catch block) and other cases + happy paths handled here.